zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumCnxManager.java

```diff
@@ -957,8 +957,13 @@ public void run() {
                                 new ListenerHandler(address, self.shouldUsePortUnification(), self.isSslQuorum(), latch))
                         .collect(Collectors.toList());
 
-                ExecutorService executor = Executors.newFixedThreadPool(addresses.size());
-                listenerHandlers.forEach(executor::submit);
+                final ExecutorService executor = Executors.newFixedThreadPool(addresses.size());
```

Comment:
what about creating a shared ExecutorService, as a field in this class ?
this way it will be clearer the life cycle of this pool.

also, probably it would be worth to not use a pool in case of a list of addresses of size 1, but this may be a follow up work

Reply:
> what about creating a shared ExecutorService, as a field in this class ?

I was thinking the same, but then I've found https://github.com/franz1981/zookeeper/blob/b4f9aab099880ba8ef08eaff697debe6cdeae057/zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Leader.java#L456-L468 and see that other parts in the ZK code just do it right without any hook to check for Thread "liveness".

I believe having a shared pluggable executor service could worth for embedded use cases, where people would like to keep control on thread creation, instead.

Reply:
> what about creating a shared ExecutorService, as a field in this class ?
> this way it will be clearer the life cycle of this pool.
> 
> also, probably it would be worth to not use a pool in case of a list of addresses of size 1, but this may be a follow up work

we can use a global shared thread pool just as what we do in spring. For example we use @component annotation to register a pool. 
however there's a question we need figure it out if the pool is io computed pool or a  cpu computed pool. 
I think that is io computed pool because we use to accept ticket there. 
@eolivelli

Reply:
let's stick with this patch.
we can improve things later